File: src/main/java/com/dotgoing/utils/j/option/None.java
package com.dotgoing.utils.j.option;

import java.util.function.Function;

public final class None<T> extends Option<T> {

    public None() {
    }

    @Override
    public <R> Option<R> map(Function<? super T, ? extends R> transformer) {
        return new None<>();
    }

    @Override
    public <R> Option<R> flatMap(Function<? super T, ? extends Option<? extends R>> transformer) {
        return new None<>();
    }

    @Override
    public T value() {
        throw new RuntimeException("none has no value");
    }

    @Override
    public boolean hasValue() {
        return false;
    }

    @Override
    public String toString() {
        return "None()";
    }
}

File: src/main/java/com/dotgoing/utils/core/option/None.kt
package com.dotgoing.utils.core.option

class None<out T>(val exception: Exception = OptionException()) : Option<T>() {

    constructor(string: String) : this(OptionException(string))

    override fun hasValue(): Boolean {
        return false
    }

    override fun value(): T {
        throw OptionStateException("none obj has no value : ${this}")
    }

    override fun errorCode(): Int {
        if (exception is OptionException) {
            return exception.code
        }
        val default = OptionException()
        return default.code
    }

    override fun error(): Exception {
        return exception

    }

    override fun toString(): String {
        return exception.message ?: "none error"
    }
}